File: visitor/cpp/visitor_test.cc
#include <gtest/gtest.h>
#include "visitor/cpp/visitor.h"

std::string captureStdout(std::function<void()> func) {
    std::stringstream buffer;
    std::streambuf* old = std::cout.rdbuf(buffer.rdbuf());

    func();

    std::cout.rdbuf(old);
    return buffer.str();
}

TEST(VisitorTest, ServiceRequestVisitor) {
    CustomerCol c;
    c.add(new EnterpriseCustomer("A company"));
    c.add(new EnterpriseCustomer("B company"));
    c.add(new IndividualCustomer("bob"));

    std::string rsp = captureStdout([&]() {
        c.accept(new ServiceRequestVisitor());
    });

    std::string expect =
        "serving enterprise customer A company\n"
        "serving enterprise customer B company\n"
        "serving individual customer bob\n";

    EXPECT_EQ(rsp, expect);
}

TEST(VisitorTest, AnalysisVisitor) {
    CustomerCol c;
    c.add(new EnterpriseCustomer("A company"));
    c.add(new IndividualCustomer("bob"));
    c.add(new EnterpriseCustomer("B company"));

    std::string rsp = captureStdout([&]() {
        c.accept(new AnalysisVisitor());
    });

    std::string expect =
        "analysis enterprise customer A company\n"
        "analysis enterprise customer B company\n";

    EXPECT_EQ(rsp, expect);
}

File: visitor/visitor_test.go
package visitor

import (
	"bytes"
	"fmt"
	"io"
	"os"
	"testing"
)

func captureStdout(f func()) string {
	old := os.Stdout
	r, w, _ := os.Pipe()
	os.Stdout = w

	f()

	w.Close()
	os.Stdout = old

	var buf bytes.Buffer
	io.Copy(&buf, r)

	return buf.String()
}

func TestRequestVisitor(t *testing.T) {
	c := &CustomerCol{}
	c.Add(NewEnterpriseCustomer("A company"))
	c.Add(NewEnterpriseCustomer("B company"))
	c.Add(NewIndividualCustomer("bob"))
	rsp := captureStdout(func() { c.Accept(&ServiceRequestVisitor{}) })
	expectRsp := (`serving enterprise customer A company
serving enterprise customer B company
serving individual customer bob
`)
	if rsp != expectRsp {
		fmt.Println(expectRsp)
		fmt.Println(rsp)
		t.Fail()
	}
}

func TestAnalysis(t *testing.T) {
	c := &CustomerCol{}
	c.Add(NewEnterpriseCustomer("A company"))
	c.Add(NewIndividualCustomer("bob"))
	c.Add(NewEnterpriseCustomer("B company"))
	rsp := captureStdout(func() { c.Accept(&AnalysisVisitor{}) })
	expectRsp := (`analysis enterprise customer A company
analysis enterprise customer B company
`)
	if rsp != expectRsp {
		fmt.Println(len(expectRsp))
		fmt.Println(len(rsp))
		t.Fail()
	}
}
